lib/puttyhosts/puttyhosts_test.go

```diff
@@ -305,3 +309,132 @@ func TestNaivelyValidateHostname(t *testing.T) {
+		input         string
+		desiredOutput []string
+		shouldError   bool
+	}{
+		{
+			input: "*.foo.example.com || *.bar.example.com",
+			desiredOutput: []string{
+				"*.foo.example.com",
+				"*.bar.example.com",
+			},
+			shouldError: false,
+		},
+		{
+			input:         "test",
+			desiredOutput: []string{"test"},
+			shouldError:   false,
+		},
+		{
+			input: "*.example.com || test || teleport.test.com",
+			desiredOutput: []string{
+				"*.example.com",
+				"test",
+				"teleport.test.com",
+			},
+			shouldError: false,
+		},
+		{
+			input: "*.example.com || test || teleport.test.com || longstring || *.wow.com",
+			desiredOutput: []string{
+				"*.example.com",
+				"test",
+				"teleport.test.com",
+				"longstring",
+				"*.wow.com",
+			},
+			shouldError: false,
+		},
+		{
+			input:         "*.example.com && port:22",
+			desiredOutput: []string(nil),
+			shouldError:   true,
+		},
+		{
+			input:         "*.example.com && ! *.extrasecure.example.com",
+			desiredOutput: []string(nil),
+			shouldError:   true,
+		},
+		{
+			input:         "(*.foo.example.com || *.bar.example.com) && port:0-1023",
+			desiredOutput: []string(nil),
+			shouldError:   true,
+		},
+		{
+			input:         "(*.foo.example.com || *.bar.example.com || *.qux.example.com) && port:0-1023",
+			desiredOutput: []string(nil),
+			shouldError:   true,
+		},
+		{
+			input:         "((*.foo.example.com || *.bar.example.com || *.qux.example.com) && port:0-1023) || teleport.example.com",
+			desiredOutput: []string(nil),
+			shouldError:   true,
+		},
+		{
+			input:         "((*.example.com || lol.example.com && port:22) && port:1024) || test.teleport.com || teleport.test.com",
+			desiredOutput: []string(nil),
+			shouldError:   true,
+		},
+		{
+			input:         "*.example.com || lol.example.com | test.teleport.com",
+			desiredOutput: []string(nil),
+			shouldError:   true,
+		},
+		{
+			input:         "*.example.com && lol.example.com || test.teleport.com",
+			desiredOutput: []string(nil),
+			shouldError:   true,
+		},
+		{
+			input:         "*.example.com || lol.example.com || | test.teleport.com",
+			desiredOutput: []string(nil),
+			shouldError:   true,
+		},
+		{
+			input:         "*.example.com || lol.example.com || | test.teleport.com || ",
+			desiredOutput: []string(nil),
+			shouldError:   true,
+		},
+		{
+			input:         "*.example.com || lol.example.com || test.teleport.com ",
+			desiredOutput: []string(nil),
+			shouldError:   true,
+		},
+		{
+			input:         " *.example.com || lol.example.com",
+			desiredOutput: []string(nil),
+			shouldError:   true,
+		},
+		{
+			input:         "*.example.com || lol.example.com || test.teleport.com || \"\"",
+			desiredOutput: []string(nil),
+			shouldError:   true,
+		},
+		{
+			input:         "*.example.com || lol.example.com || .",
+			desiredOutput: []string(nil),
+			shouldError:   true,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.input, func(t *testing.T) {
```

Comment:
Sorry it's tedious but could we give each of the cases a name? It will should up nicer in the `go test` output and it serves as a comment, on some of the cases it isn't completely obvious why they should be errors
```suggestion
		t.Run(tt.name, func(t *testing.T) {
```